### DISCUSSION TITLE
Is there a way of using APIs (Java)  to read in an smt2 file ?

### DISCUSSION THREAD

**Discussion by classicwuhao** [cvc5/cvc5#8345] [Q&A]
I am wondering is it possible to use APIs to read in an smt2 file for solving ?

**Comment by 4tXJ7f** [comment]
Unfortunately, we currently do not have a user-facing API for parsing (neither for C++ nor Java). We would like to support this in the future (together with a new parser), but we don't have a concrete timeline right now. Out of curiosity, what is your use case?

**Reply by classicwuhao** [reply]
We have multiple running research projects using SMT solvers, already include CVC5 for one of the projects but would like to include CVC5 for another project. However, the other one uses our own built-in engine to generate smt2 file and sends it to the solver for verification.  It will be nice if CVC5 provides a way of parsing an SMT file as input and have some model retrieval APIs support. So we can just connect CVC5 into our own engine as another back-end solver.

**Comment by kfriedberger** [comment]
Is there any update on this topic yet?
The [JavaSMT](https://github.com/sosy-lab/java-smt) project would also like to have a working parser in CVC5 via the API.

Our use-case or requirement is as follows:
- The required method parses from String to term(s). The String is formatted as SMTLIB2 and contains the wanted term(s) as assertion(s). The String also includes all symbols and uninterpreted functions used in the assertion(s).
- The required method knows the context and merges the symbols that exist in the current context/solver and the symbols from the SMTLIB input, and crashes/warns on conflicts, e.g., symbols with same name and distinct types.
- We do not need the possibility to check satisfiability or query further information like models via the parser interface. Those operations are done separately after the parsing.

**Comment by classicwuhao** [comment]
Any update on APIs for reading smt2 file?

**Comment by ajreynol** [comment]
Yes, this is now available in C++ and Java, see documentation: 
https://cvc5.github.io/docs-ci/docs-main/api/cpp/inputparser.html#inputparser (C++) 
https://cvc5.github.io/docs-ci/docs-main/api/java/io/github/cvc5/InputParser.html (Java)

Support for Python is being added in this PR, which should be merged today: https://github.com/cvc5/cvc5/pull/10222

**Reply by classicwuhao** [reply]
thanks !!!

**Comment by classicwuhao** [comment]
@ajreynol I have been reading API docs and exploring examples provided. A quick question, I used InputParser to parse the smt2 file and solver to check sat(it returns sat). Now, I would like to get value for each const just like command `get-value (x1, ... ,xn)`. Is there any API that allows me to get the values?

**Comment by ajreynol** [comment]
Yes, the `Solver` has an interface for `getValue` https://github.com/cvc5/cvc5/blob/main/include/cvc5/cvc5.h#L4621. You need to enable the option `produceModels` to use this method.

**Reply by classicwuhao** [reply]
@ajreynol Thanks a lot. I think my problem is retrieving all the terms from parser or solver. I have looked at API docs and examples about `Solver` and `InputParser`. But how do I get all the terms in order to use `solver.getValue(t)` ? I have a piece of code below 
```
Solver solver = new Solver();
solver.setOption("produce-models", "true");
solver.setLogic("ALL");
SymbolManager sm = new SymbolManager(solver);
InputParser parser = new InputParser(solver,sm);
parser.setFileInput(InputLanguage.SMT_LIB_2_6,"./test.smt2");
Term t;
	do{
        	t = parser.nextTerm();
		System.out.println(t);			
	}while (!t.isNull());

Result r = solver.checkSat();
System.out.println("result:"+r);
```
my test.smt2 is as follows: 
```
(declare-fun x() Int)
(declare-fun t1()  Int)
(declare-fun t2() Int)

(assert (= x 0))
```
How to get value of `x`?

**Comment by ajreynol** [comment]
One way to do this is to initialize the solver with `freshDeclarations` to false (also available on command line with `--no-fresh-declarations`).

This will cause the parser to use the `declareFun` method will `fresh` set to `false`, see https://github.com/cvc5/cvc5/blob/main/include/cvc5/cvc5.h#L4213.

Afterwards, calling `Solver::declareFun` on your end again with `fresh=false` and *the same type and name* will make the solver return the same term as the one that was parsed.

Then, `getValue` should work as expected.


Otherwise, we may expand our interface for commands so that you would be able to access `x` as it is parsed.

**Reply by classicwuhao** [reply]
@ajreynol many thanks. I think it would be better if you guys could offer some interface to users so that they could have access to a particular variable or function, etc.

**Reply by classicwuhao** [reply]
@ajreynol I am trying to do as you suggested. For a variable `declare-fun x() Int` in my smt2 file, I do the following:
````
Sort integer = solver.getIntegerSort();
Sort[] sorts = new Sort[1];
sorts[0]=solver.getNullSort();
Term t = solver.declareFun("x",sorts,integer,false);
...
//parser reads each term as usual and gets value after solving
````
Two issues:

0. `getNullSort()` cannot be found by java. I checked Solver.java it does not contain this method even it is contained in the java API document (it may need an update?)
1. do you guys have a fully worked example? If not, I am happy to write one but before that I need some help from the code above.

**Reply by mudathirmahgoub** [reply]
In the declaration you have `declare-fun x() Int` x is a function that does not receive any argument, so you don't need to add any sort as an argument. You can try 
```java
Term x = solver.declareFun("x", new Sort[0], solver.getIntegerSort());
```
For such functions, it is just easier to use `mkConst` as follows:
```java
 Term x = solver.mkConst(solver.getIntegerSort(), "x");
```
For a null Sort or a null Term, they were associated with the solver class in the past, but they were removed.  
I checked the latest documentation, and could not find `getNullSort()` here
https://cvc5.github.io/docs-ci/docs-main/api/java/io/github/cvc5/Solver.html
Anyway, if you really need it, you would write:
```java
Sort nullSort = new Sort();
```

**Reply by mudathirmahgoub** [reply]
For examples, see these two:
https://cvc5.github.io/docs-ci/docs-main/api/java/quickstart.html
https://cvc5.github.io/docs-ci/docs-main/examples/parser.html

**Reply by classicwuhao** [reply]
Thanks. We have already gone through these examples before. What we really want is: we would like to use cvc5's APIs to parse already generated smt2 files, solve them and get value for each variable (if there is a model).

**Comment by ajreynol** [comment]
Can you try this example?
```

import io.github.cvc5.*;
import io.github.cvc5.modes.*;
public class Parser
{
  public static void main(String args[]) throws CVC5ApiException
  {
    Solver slv = new Solver();

    // set fresh declarations to false so that there is e.g. exactly one
    // declared function for ("a", Int) below.
    slv.setOption("fresh-declarations", "false");

    // construct an input parser associated the solver above
    InputParser parser = new InputParser(slv);

    String ss = "";
    ss += "(set-logic QF_LIA)\n";
    ss += "(declare-fun a () Int)\n";
    ss += "(assert (> a 4))\n";
    ss += "(assert (< a 6))\n";
    ss += "(check-sat)\n";
    parser.setStringInput(InputLanguage.SMT_LIB_2_6, ss, "MyStream");

    // get the symbol manager of the parser, used when invoking commands below
    SymbolManager sm = parser.getSymbolManager();

    // parse commands until finished
    Command cmd;
    while (true)
    {
      cmd = parser.nextCommand();
      if (cmd.isNull())
      {
        break;
      }
      cmd.invoke(slv, sm);
    }
    System.out.println("Finished parsing commands");

    // use fresh-declarations = false, so that the declared function is the
    // same as the one parsed above.
    Term a = slv.declareFun("a", new Sort[0], solver.getIntegerSort(), false);
    Term v = slv.getValue(a);
    System.out.println("expected: 5");
    System.out.println("result: " + v);
  }
}
```
Again, I think it should suffice to set `fresh-declarations` to false for the solver option and then to access the variable using `false` for the 4th arg to `Solver::declareFunction`.  Let me know if that doesn't work.

**Reply by classicwuhao** [reply]
@ajreynol Thanks a lot for this example. 👍 It didn't compile but it is suffice enough to see where went wrong in my previous example. Now, retrieving values works for me. Here is this full example:
````
Solver slv = new Solver();
slv.setOption("produce-models", "true");
slv.setOption("fresh-declarations","false");
slv.setLogic("ALL");
SymbolManager sm = new SymbolManager(slv);
InputParser parser = new InputParser(slv, sm);
parser.setFileInput(InputLanguage.SMT_LIB_2_6,"./test.smt2");
	
Command cmd;
while (!(cmd=parser.nextCommand()).isNull()) cmd.invoke(slv,sm);
   
Term x = slv.declareFun("x",new Sort[0],slv.getIntegerSort(),false);
Term t1 = slv.declareFun("t1",new Sort[0],slv.getIntegerSort(),false);
Term t2 = slv.declareFun("t2",new Sort[0],slv.getIntegerSort(),false);
   
System.out.println(slv.checkSat());
System.out.println("x:"+slv.getValue(x));
System.out.println("t1:"+slv.getValue(t1));
System.out.println("t2:"+slv.getValue(t2));
````
test.smt2 file is as follows:
````
(declare-fun x() Int)
(declare-fun t1()  Int)
(declare-fun t2() Int)

(assert (> x 5))
(assert (> (+ t1 t2) x))
````
Output:
````
sat
x:6
t1:7
t2:0
````

**Comment by classicwuhao** [comment]
@ajreynol Since cvc5 has been upgrade to 1.2.1, can you shed light on how to get the above example working? Many methods are deprecated, we can no longer retrieve value for each term through above approach. It would be better if we could have a fully worked example with java APIs. So every time an update to the APIs, this example can also gets updated.
